src/Generator.cpp

```diff
@@ -21,362 +26,655 @@
-      for (const auto& id_energy : event.hits()) { 
-        hCellEnergyEvtMap.at(id_energy.first).Reset();
-        hCellEnergyEvtMap.at(id_energy.first).Fill(id_energy.second);
-      }
-      canvas.emplace_back(display(hCellEnergyEvtMap,iev));    
+    // start main loop on all events
+    double hit_outside_geom = 0.;
+    double tot_hits = 0;
 
+    // Histograms for transverse profile
+    TH1F *hTransverseProfile[layer_max-layer_min];
+    char* hTransProfName = new char[30];
+
+    for (int i = layer_min; i < layer_max; i++) {
+        sprintf(hTransProfName, "hTransverseProfile_%d", i);
+        hTransverseProfile[i] = new TH1F(hTransProfName, "Generated transverse profile (cm)", 600, 0., 120);
     }
-    output_.fillTree(event, geometry_);
 
-  }
-  std::cout << std::endl;
-  std::cout << nevents << " events generated " << std::endl;  
-  std::cout << std::endl;
+    // Build the tree map (except for external geometry) ; one key corresponds to one layer
+    std::unordered_map<int, Tree*> tree_map;
+
+    if (parameters_.geometry().type!=Parameters::Geometry::Type::External) {
+
+        for (int layer_id = layer_min; layer_id < layer_max; layer_id++) {
+
+            std::vector<double > cell_x;
+            std::vector<double > cell_y;
+            for (Cell c : cell_collection[layer_id]) {
+                cell_x.push_back(c.getX());
+                cell_y.push_back(c.getY());
+            }
+
+            double x_min = *std::minmax_element(cell_x.begin(), cell_x.end()).first;
+            double x_max = *std::minmax_element(cell_x.begin(), cell_x.end()).second;
+            double y_min = *std::minmax_element(cell_y.begin(), cell_y.end()).first;
+            double y_max = *std::minmax_element(cell_y.begin(), cell_y.end()).second;
+
+            // Geometry build with the cell center position, add a margin for the
+            // square corners containing the boarder cells
+            x_min -= parameters_.geometry().small_cell_side;
+            x_max += parameters_.geometry().large_cell_side;
+            y_min -= parameters_.geometry().large_cell_side * 2;
+            y_max += parameters_.geometry().large_cell_side * 2;
+
+            Rectangle* plan = new Rectangle(
+                new Point((float)x_min, (float)y_max),
+                new Point((float)x_max, (float)y_min)
+            );
+
+            Tree* tree = new Tree(plan, 5);
+            float x_c, y_c;
+            double side;
+            for(Cell& c : cell_collection[layer_id]) {
+
+                x_c = float(c.getX());
+                y_c = float(c.getY());
+                float cell_radius = sqrt((x_c*x_c)+(y_c-y_c));
+
+                if (cell_radius <= float(parameters_.geometry().limit_first_zone))
+                    side = parameters_.geometry().small_cell_side;
+                else
+                    side = parameters_.geometry().large_cell_side;
+
+                Point* cornerA = new Point(x_c - float(side*sqrt(3)/2), y_c + float(side));
```

Comment:
It's generally better to use smart pointers (e.g. `unique_ptr`) to avoid memory problems, or eventually no pointer at all. Here for instance the Point objects don't seem to be deleted, which would create memory leaks.

Reply:
Fixed